### PR TITLE
feat: add `maybe_reset_index` for pandas-like dataframe or series

### LINKED ISSUE #1070: [Enh]: add `maybe_reset_index`
### Please describe the purpose of the new feature or describe the problem to solve.

We have `maybe_get_index` and `maybe_set_index`, yet `maybe_reset_index` is missing.

### Suggest a solution if possible.

Implementation should look something along the following lines:

```py
def maybe_reset_index(df, *, drop: bool):
    native_frame = to_native(df)
    if is_pandas_like_dataframe(native_frame):
        return from_native(native_frame.reset_index(drop=drop))
    return df
```

### If you have tried alternatives, please describe them below.

_No response_

### Additional information that may help us understand your needs.

_No response_

### PR DESCRIPTION
**Pull request by raisadz**
## What type of PR is this? (check all applicable)

- [ ] 💾 Refactor
- [x] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [ ] 📝 Documentation
- [ ] ✅ Test
- [ ] 🐳 Other

## Related issues 

- Related issue # 
- Closes #1070

## Checklist

- [x] Code follows style guide (ruff)
- [x] Tests added 
- [x] Documented the changes

### CODE REVIEW

File: narwhals/utils.py [hunk 1]
```diff
@@ -307,6 +307,39 @@ def maybe_set_index(df: T, column_names: str | list[str]) -> T:
...
+        1  2  5
+        >>> series_pd = pd.Series([1, 2])
+        >>> series = nw.from_native(series_pd, series_only=True)
+        >>> nw.maybe_get_index(series)
+        RangeIndex(start=0, stop=2, step=1)
+    """
+    obj_any = cast(Any, obj)
+    native_obj = to_native(obj_any)
+    if is_pandas_like_dataframe(native_obj) or is_pandas_like_series(native_obj):
+        return native_obj.reset_index(drop=True)
```
Comment (FBruzzesi) [inline narwhals/utils.py:339]: I think this should return a narwhals DataFrame/Series back:

```suggestion
        return from_native(
            native_obj.reset_index(drop=True), allow_series=True,
        )
```

**Review by FBruzzesi — COMMENTED** [review]
Hey @raisadz thanks for addressing this! It definitly was in my wish list 🚀

I have a tiny comment regarding the return object 🙈

**Comment by raisadz** [comment]
> Hey @raisadz thanks for addressing this! It definitly was in my wish list 🚀
> 
> I have a tiny comment regarding the return object 🙈

Hey @FBruzzesi, thanks for noting that the return should be a narwhals object, I addressed it

File: narwhals/stable/v1/__init__.py [hunk 2]
```diff
@@ -1796,6 +1797,35 @@ def maybe_set_index(df: T, column_names: str | list[str]) -> T:
     return nw_maybe_set_index(df, column_names)
 
 
+def maybe_reset_index(obj: T) -> Any:
```
Comment (MarcoGorelli) [inline narwhals/stable/v1/__init__.py:1800]: the return type should also be `T` (i.e. it's the same type of the input)

**Review by MarcoGorelli — COMMENTED** [review]
thanks @raisadz ! just a small comment

File: narwhals/utils.py [hunk 3]
```diff
@@ -307,6 +308,42 @@ def maybe_set_index(df: T, column_names: str | list[str]) -> T:
     return df_any  # type: ignore[no-any-return]
 
 
+def maybe_reset_index(obj: T) -> T:
+    """
+    Reset the index to the default integer index of a DataFrame or a Series, if it's pandas-like.
+
+    Notes:
+        This is only really intended for backwards-compatibility purposes,
+        for example if your library already aligns indices for users.
```
Comment (FBruzzesi) [inline narwhals/utils.py:317]:
```suggestion
        for example if your library already reset the index for users.
```

File: narwhals/stable/v1/__init__.py [hunk 4]
```diff
@@ -1796,6 +1797,35 @@ def maybe_set_index(df: T, column_names: str | list[str]) -> T:
     return nw_maybe_set_index(df, column_names)
 
 
+def maybe_reset_index(obj: T) -> T:
+    """
+    Reset the index to the default integer index of a DataFrame or a Series, if it's pandas-like.
+
+    Notes:
+        This is only really intended for backwards-compatibility purposes,
+        for example if your library already aligns indices for users.
```
Comment (FBruzzesi) [inline narwhals/stable/v1/__init__.py:1806]:
```suggestion
        for example if your library already reset the index for users.
```

**Review by FBruzzesi — COMMENTED** [review]
Found another tiny misalignment with the functionality, then I think it is good to go 👌

File: narwhals/utils.py (hunk 3 above)
Comment (MarcoGorelli) [reply narwhals/utils.py:317]: *resets 😉

Comment (FBruzzesi) [reply narwhals/utils.py:317]: Fixed! Thank you :)

File: tests/utils_test.py [hunk 5]
```diff
@@ -84,6 +84,30 @@ def test_maybe_get_index_polars() -> None:
...
+    assert_series_equal(nw.to_native(result_s), expected_s)
+
+
+def test_maybe_reset_index_polars() -> None:
+    df = nw.from_native(pl.DataFrame({"a": [1, 2, 3], "b": [4, 5, 6]}))
+    result = nw.maybe_reset_index(df)
+    assert result is df
+    series = nw.from_native(pl.Series([1, 2, 3]), series_only=True)
+    result = nw.maybe_reset_index(series)
+    assert result is series
```
Comment (MarcoGorelli) [inline tests/utils_test.py:108]:
```suggestion
    result_s = nw.maybe_reset_index(series)
    assert result_s is series
```

Comment (raisadz) [reply tests/utils_test.py]: Thanks for fixing this @MarcoGorelli! I wasn't sure why `mypy` in CI  was failing there

**Pull request merged**